### PR TITLE
DT-601 service for court hearings

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
Initial PR for scheduling of court hearings (initially for PECS).

Note the service is not wired into to any endpoint hence it is a work in progress.

### CODE REVIEW

File: src/test/java/net/syscon/elite/repository/jpa/repository/CourtEventRepositoryTest.java [hunk 1]
```diff
@@ -39,8 +39,7 @@
     private TestEntityManager entityManager;
 
     @Test
-    public void court_event_can_be_saved_and_retrieved() {
-        var id = 999L;
```
Comment (matt-crowson-moj) [inline src/test/java/net/syscon/elite/repository/jpa/repository/CourtEventRepositoryTest.java]: This is a key change to note here.  The id is now populated via a sequence.

File: src/main/java/net/syscon/elite/repository/jpa/model/CourtEvent.java [hunk 2]
```diff
@@ -34,6 +36,8 @@
 @ToString(exclude = {"offenderBooking", "offenderCourtCase"})
 public class CourtEvent extends AuditableEntity {
 
+    @SequenceGenerator(name = "EVENT_ID", sequenceName = "EVENT_ID", allocationSize = 1)
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/repository/jpa/model/CourtEvent.java]: Changed here to pull id from sequence on the DB (as per NOMIS).

File: src/main/java/net/syscon/elite/service/CourtHearingsService.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    public CourtHearingsService(final OffenderBookingRepository offenderBookingRepository,
+                                final CourtEventRepository courtEventRepository,
+                                final AgencyLocationRepository agencyLocationRepository) {
+        this.offenderBookingRepository = offenderBookingRepository;
+        this.courtEventRepository = courtEventRepository;
+        this.agencyLocationRepository = agencyLocationRepository;
+    }
+
+    @Transactional
+    public CourtHearing scheduleHearing(final Long bookingId, final PrisonToCourtHearing hearing) {
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/service/CourtHearingsService.java]: Hopefully quite obvious but this is a work in progress.  Not used outside of service layer though.

File: src/main/java/net/syscon/elite/service/CourtHearingsService.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+        // TODO throw entity not found exception (check is prison?)
+        AgencyLocation fromPrison = agencyLocationRepository.findById(hearing.getFromPrisonLocation()).orElseThrow();
+
+        // TODO throw entity not found exception (check is court)
+        AgencyLocation toCourt = agencyLocationRepository.findById(hearing.getToCourtLocation()).orElseThrow();
+
+        // TODO sort out reference codes!!!
+        CourtEvent courtEvent = CourtEvent.builder()
+                .courtEventType(new EventType("CRT", "Court Action"))
```
Comment (Mjwillis) [inline src/main/java/net/syscon/elite/service/CourtHearingsService.java]: EventType is reference data? If so - should this be loaded?

File: src/main/java/net/syscon/elite/service/CourtHearingsService.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                .courtLocation(toCourt)
+                .directionCode("OUT")
+                .eventDate(hearing.getCourtHearingDateTime().toLocalDate())
+                .eventStatus(new EventStatus("SCH", "Scheduled (Approved)"))
```
Comment (Mjwillis) [inline src/main/java/net/syscon/elite/service/CourtHearingsService.java]: Also here the EventStatus?

File: src/main/java/net/syscon/elite/service/CourtHearingsService.java (hunk 4 above)
Comment (matt-crowson-moj) [reply src/main/java/net/syscon/elite/service/CourtHearingsService.java]: Yes.  I have noted this in the TODO above.  The current code base isn't really set up quite right  for retrieval.  Something I will look at resolving today.

File: src/main/java/net/syscon/elite/service/CourtHearingsService.java (hunk 5 above)
Comment (matt-crowson-moj) [reply src/main/java/net/syscon/elite/service/CourtHearingsService.java]: Same as other comment.  On TODO.

**Pull request merged**